Replace remote Enrich hack with proper handling of remote Enrich

smalyshev opened the pull request:
- The remote Enrich hack and the verification workaround that it necessitates are removed
- Limit with remote Enrich is handled by duplicating the limits around Enrich
- TopN with remote Enrich is handled by duplicating around Enrich and adjusting projections
- Introduced `CardinalityPreserving` interface to mark nodes that preserve cardinality
- Introduced the capability to run optimizations only on coordinator plan (`CoordinatorOnly` interface)
- Added post-optimizer check that ensures remote Enrich is not moved to the coordinator plan

TODO: this does not fix the issue of local (not remote) Lookup Join always being forced on the coordinator by Limits, we may address this in the followup (https://github.com/elastic/elasticsearch/issues/136227), this one is meaty enough as it is. 

Closes https://github.com/elastic/elasticsearch/issues/115897
Fixes https://github.com/elastic/elasticsearch/issues/118531

smalyshev committed:
Add check that remote enrich stays on remote side

smalyshev committed:
Merge branch 'main' into remote-enrich-check

smalyshev committed:
Make it more generic

Commit:
[CI] Auto commit changes from spotless

smalyshev committed:
Verifier refactor

smalyshev committed:
Merge branch 'main' into remote-enrich-check

smalyshev committed:
Merge branch 'main' into remote-enrich-check

smalyshev committed:
Fix the fix for https://github.com/elastic/elasticsearch/issues/118531

smalyshev committed:
Merge branch 'main' into remote-enrich-check

smalyshev committed:
Remove old hacks

smalyshev committed:
Replace remote Enrich hack with proper handling of remote Enrich

smalyshev committed:
Don't stop for ANY enrich

smalyshev committed:
Add TopN handling

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
Oops, forgot the file

smalyshev committed:
fix multiple topNs

smalyshev committed:
fix tests

smalyshev committed:
test fixes

smalyshev committed:
Ban Limit + MvExpand before remote Enrich

smalyshev committed:
Update docs/changelog/135051.yaml

smalyshev committed:
spotless

smalyshev committed:
Move check pre optimizer

smalyshev committed:
Merge branch 'remote-enrich-limit' into entich-remote-unhack

smalyshev committed:
fix tests

Commit:
[CI] Update transport version definitions

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichLimit.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    protected LogicalPlan rule(Enrich en, LogicalOptimizerContext ctx) {
+        if (en.mode() == Enrich.Mode.REMOTE) {
+            Set<Limit> seenLimits = new HashSet<>();
```
Maybe better to use an identity hash set?

alex-spies reviewed (commented):
I had a first look at the new optimizer rules and left some comments.

smalyshev committed:
test fixes

Commit:
[CI] Update transport version definitions

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
Update x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/TopN.java

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

smalyshev committed:
fix merge

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java:
```diff
@@ -0,0 +1,93 @@
+                        } else {
+                            // No need for aliasing - then it's simple, just copy the TopN on top and mark the original as local
+                            LogicalPlan transformedEnrich = en.transformDown(TopN.class, t -> t == top ? top.withLocal(true) : t);
+                            return new TopN(top.source(), transformedEnrich, top.order(), top.limit());
```
It should handle them on repeated application of the rule.

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java:
```diff
@@ -0,0 +1,93 @@
+                    || plan instanceof PipelineBreaker) {
+                    break;
+                }
+                if (plan instanceof UnaryPlan u) {
```
How would you define them? Probably if it's not safe then the plan in general wouldn't be executable?

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java:
```diff
@@ -0,0 +1,93 @@
+                        plan = top.child();
+                        continue;
+                    } else {
+                        // Create a fake OrderBy and "push" Enrich through it to generate aliases
```
I'm not sure I understood the last part... We do want limits before Enrich, don't we? Otherwise we may be enriching many more records than we need to?

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java:
```diff
@@ -0,0 +1,93 @@
+                            OrderBy order = (OrderBy) proj.child();
+                            Enrich enrich = (Enrich) order.child();
+                            Eval eval = (Eval) enrich.child();
+                            // We insert the evals above the original TopN, so that the copy TopN works on the renamed fields
```
Yes, I didn't deal with intermediate projections yet, I'll look into it soon.

smalyshev committed:
test fixes

smalyshev committed:
spotless

smalyshev committed:
Fix tests

smalyshev committed:
Cleanup

smalyshev committed:
Fix test

Commit:
[CI] Auto commit changes from spotless

smalyshev committed:
Add project handling

Commit:
[CI] Update transport version definitions

smalyshev committed:
Add two topN test

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
Clean up plan/optimizer structure

smalyshev committed:
Introduce CardinalityExpanding

smalyshev committed:
Switch to CardinalityPreserving

smalyshev committed:
Block multiple TopN handling as it's semantically unsound

smalyshev committed:
Remove todo

breskeby committed:
[Build] Update checkstyle from 10.3 to 11.0.1

This should fix a asm incompatibility we have seen in our builds

smalyshev committed:
Merge branch 'pr/135381' into entich-remote-unhack

smalyshev committed:
Remove exclusion

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

breskeby committed:
Add missing dependency verification

smalyshev committed:
Merge branch 'pr/135381' into entich-remote-unhack

breskeby committed:
[Build] update eclipse formatter used by spotless

We cannot directly declare the ecl and jdt core dependency but we can tweak the used
dependency by spotless in our precommit plugin

smalyshev committed:
Merge branch 'pr/135382' into entich-remote-unhack

smalyshev committed:
spotless

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
cleanup

smalyshev committed:
Cleanups

smalyshev committed:
clean test

smalyshev committed:
moar tests

smalyshev committed:
oops typo

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Limit.java:
```diff
@@ -11,14 +11,15 @@
 import java.util.Objects;
 
-public class Limit extends UnaryPlan implements TelemetryAware, PipelineBreaker {
+public class Limit extends UnaryPlan implements TelemetryAware, PipelineBreaker, ExecutesOn {
```
It's a little confusing now that this is `PipelineBreaker` but can still execute on `ANY` sometimes.

Out of scope, but maybe we'll want to reconcile the two interfaces later; `PipelineBreaker` could be another enum variant; so we'd have "executes anywhere", "executes only on data nodes", "breaks the pipeline if we're not already on the coordinator", "runs only on the coordinator".

smalyshev replied:
The thing is, commands like LIMIT are "sometimes pipeline breakers" now, but we don't have an interface like that. It may be a good idea to think about how to refine that, but I don't think in this patch.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/mapper/Mapper.java:
```diff
@@ -170,7 +121,13 @@ private PhysicalPlan mapUnary(UnaryPlan unary) {
 
         if (unary instanceof Limit limit) {
-            mappedChild = addExchangeForFragment(limit, mappedChild);
+            // Avoid extra limit for remote Enrich, since it has already been applied before Enrich if it's duplicate.
```
The extra limit is fine. It'll be optimized away on the data node. (limits are pushed past enrichs.) I'd leave this the data node's responsibility, otherwise the code might get brittle.

(We have no way to prove that this duplicated limit isn't somehow still crucial. Other optimizations may have picked up the duplicate limit and done something with it. In fact, it's not incorrect for other optimizations to have removed the duplicate limit before we reach here.)

smalyshev replied:
In the case of Enrich, it's not crucial because we will have a copy of it pre-enrich. It's not true for other operations (like joins) because they can expand cardinality. So the mapper does the right thing in the general case, but in specifically cause of Enrich this limit is useless. I mainly inserted this to not have to redo a bunch of tests by inserting extra limit, but if you think it's fine we could leave it alone and let the data node optimizer handle it (I'll check if it does).

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/plan/logical/TopNSerializationTests.java:
```diff
@@ -40,16 +41,26 @@ protected TopN mutateInstance(TopN instance) throws IOException {
+    @Override
+    protected TopN copyInstance(TopN instance, TransportVersion version) throws IOException {
+        // TopN#local is ALWAYS false after serialization.
+        TopN deserializedCopy = super.copyInstance(instance.withLocal(false), version);
```
I don't think you need to set the original instances `local`, it'll be igored by the serialization, anyway.

alex-spies reviewed (commented):
Done a first pass - need another one. I'm like halfway through it, but haven't fully thought through the added optimizer rules, yet.

I noticed that this doesn't add new optimizer tests, yet. That's something I think we should really add before this can be merged; I'd go and try really hard to find edge cases where the hoisting mechanism, the new local-only topn/limit etc. may break. 

(The generative tests will also do that to some extent (let's make sure they actually create remote-only enrichs!), but it's good to have a human-made baseline.)

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/CombineLimitTopN.java:
```diff
@@ -0,0 +1,37 @@
+import org.elasticsearch.xpack.esql.plan.logical.TopN;
+
+/**
+ * Combines a Limit followed by a TopN into a single TopN.
```
Correct, this is to fix the effects of `HoistRemoteEnrichTopN`. I'll add a comment.

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/CombineLimitTopN.java:
```diff
@@ -0,0 +1,37 @@
+    @Override
+    public LogicalPlan rule(Limit limit, LogicalOptimizerContext ctx) {
+        if (limit.child() instanceof TopN topn) {
+            int thisLimitValue = (int) limit.limit().fold(ctx.foldCtx());
```
Ah, this is good. I'll change it.

smalyshev commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -7039,7 +7039,7 @@ public void testEnrichBeforeLimit() {
         }
         {
-            var plan = physicalPlan("""
+            var plan = physicalPlanNoSerializationCheck("""
```
It's because local flags in limits are not serialized (because they are pointless on remote side). I'll add a comment. I wish there were a better way to handle this (like, compare only non-transient members) but not sure how.

smalyshev commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -7229,11 +7229,14 @@ public void testEnrichAfterTopN() {
             var fragment = as(exchange.child(), FragmentExec.class);
-            var enrich = as(fragment.fragment(), Enrich.class);
+            var dupTopN = as(fragment.fragment(), TopN.class);
+            assertThat(dupTopN.limit().toString(), equalTo("10"));
```
I think `limit()` was of some weird type, not plain number, so that's why I used it. I'll take another look.

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/InlineStats.java:
```diff
@@ -38,7 +38,13 @@
+        NamedWriteable,
+        SurrogateLogicalPlan,
+        TelemetryAware,
+        CardinalityPreserving,
```
OK I can drop it, I don't think it matters for my case, just added some for completeness' sake.

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Limit.java:
```diff
@@ -75,28 +83,48 @@ public String getWriteableName() {
+        return new Limit(source(), limit, newChild, duplicated, local);
+    }
+
+    public Limit combine(Limit other, FoldContext ctx) {
```
OK, you are right, it's not used anywhere else anyway. I wanted it to be a separate method, but one for `PushDownAndCombineLimits` is better.

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -165,6 +168,7 @@ protected static Batch<LogicalPlan> substitutions() {
     protected static Batch<LogicalPlan> operators(boolean local) {
```
I made `LocalLogicalPlanOptimizer#localOperators` respect the `CoordinatorOnly` marker, so that's the same thing now, except for rule that work on both sides, but differently (`PropagateEmptyRelation`). I agree that it could be refined further but not sure the best way yet.

smalyshev committed:
Feedback, pass 1

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/CardinalityPreserving.java:
```diff
@@ -0,0 +1,30 @@
+package org.elasticsearch.xpack.esql.plan.logical;
+
+/**
+ * This interface marks a command which does not add or remove rows.
```
I meant this interface more in the "necessary" sense - i.e. if something is marked with it, it's _might be_ ok to interchange it with limit, if it's not, then definitely not. From this POV, `CHANGE_POINT` probably should be excluded, but we may exclude some other preserving ops for reasons... I am not sure if it's ok to push limit under `CHANGE_POINT` for example - that's a good question which needs to be answered, but I'll remove the marker for now until it is.

alex-spies replied:
Please, let's update the precise definition of this interface. I gave a suggestion above. This interface is implicitly "any command that we can hoist topn and limits past", but currently it says "can be commutated with limits" - which is both too much (`LIMIT X | MY_COMMAND` has to be the same as `local LIMIT X | MY_COMMAND | LIMIT X`) and too little (doesn't say anything about topns).

smalyshev committed:
More feedback

smalyshev committed:
Remove limit optimization - data node optimizer will take care of it

smalyshev committed:
moar tests

smalyshev committed:
avoid forbidden lore

smalyshev committed:
More tests

smalyshev committed:
TopN tests

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
fix test

alex-spies committed:
Refine CoordinatorOnly interface for rules

alex-spies reviewed (commented):
Ok, I spent some time mulling over the hoisting rules.

I think the limit one is nice and correct, but I'm pretty sure the one for topn is still missing some (annoying but) important cases - see below.

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java:
```diff
@@ -0,0 +1,108 @@
+                if (plan instanceof TopN top && top.local() == false) {
+                    // Create a fake OrderBy and "push" Enrich through it to generate aliases
+                    Enrich topWithEnrich = (Enrich) en.replaceChild(new OrderBy(top.source(), en.child(), top.order()));
+                    LogicalPlan pushPlan = PushDownUtils.pushGeneratingPlanPastProjectAndOrderBy(topWithEnrich);
```
Hm I guess we'd need to check intermediate nodes then and insert the eval in that case. I don't see why there wouldn't be a possibility to fix it - if we insert the eval, then the field won't go anywhere. We could go as far as _always_ insert the eval if we wanted to, I think... worst thing it'd do is to send an extra column over the wire. Not ideal, but shouldn't be a huge cost. I'll look into it more, thanks for pointing out this case!

alex-spies replied:
For the current implementors of CardinalityPreserving, I think we should be able to make it work.

As a general mechanism that will remain correct when folks add new nodes and mark them as CardinalityPreserving, I don't know if we can make that work (without making the CardinalityPreserving interface more complicated than it is). Folks might e.g. implement their own custom projections (without inheriting from Project) that will just drop everything except some fixed attributes that they want to be in the output.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownAndCombineLimits.java:
```diff
@@ -71,11 +72,26 @@ public LogicalPlan rule(Limit limit, LogicalOptimizerContext ctx) {
             // left-hand side, so adding a limit in there would lead to the right-hand side work on incomplete data.
             // To avoid repeating this infinitely, we have to set duplicated = true.
-            return duplicateLimitAsFirstGrandchild(limit);
+            // We use withLocal = false because if we have a remote join it will be forced into the fragment by the mapper anyway,
```
If you agree that this can maybe be simplified, let's ideally do that in a follow-up; this PR is tricky enough by itself :)

smalyshev committed:
Take care of shadowing for TopN

smalyshev committed:
optimize a bit

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
comments

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
Refactor HoistRemoteEnrichTopN to be more future-proof

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
fix tests

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
simplify output checking

smalyshev committed:
take outputs from the loop

smalyshev committed:
Fix attribute identity

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
Just use id for Attribute identity

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java:
```diff
@@ -0,0 +1,168 @@
+                        // Replace the missing attributes in order expressions with their aliases
+                        for (Order expr : topN.order()) {
+                            rewrittenExpressions.add(expr.transformUp(Attribute.class, attr -> {
+                                if (missingNames.contains(attr.name())) {
```
nit: comparing by name isn't wrong here, but usually we use an `AttributeSet`, where `.contains` works by NameId. That'd avoid the dance where we start with a `Set<Attribute>` (not the same!) and then create a `Set<String>` from it.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java:
```diff
@@ -0,0 +1,168 @@
+                            }));
+                        }
+                        var replacedAttributes = aliasesForReplacedAttributesBuilder.build();
+                        // We need to create a copy of attributes used by TonP, because Enrich or some command on the way overwrites them.
```
```suggestion
                        // We need to create a copy of attributes used by TopN, because Enrich or some command on the way overwrites them.
```

alex-spies reviewed (commented):
Ok, I think there's one (final) bugfix require in HoistRemoteEnrichTopN, but it's easily fixable and the hoisting looks right to me other than that - see below.

smalyshev committed:
typo

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java:
```diff
@@ -0,0 +1,127 @@
+                            case Project pr -> {
+                                List<NamedExpression> allFields = new LinkedList<>(pr.projections());
+                                allFields.addAll(eval.fields());
+                                yield pr.withProjections(allFields);
```
It's surprisingly hard to test though - I can't keep any `Project` under Enrich/TopN, they all get moved upwards. So if you have an idea for a test query, very welcome!

alex-spies replied:
Rather than testing the whole optimizer chain, you can construct a specific logical plan and test the application of a single rule pass.

Alternatively, you can invent a test-only logical plan node that implements CardinalityPreserving and place a projection under it. The projection won't travel upwards because there's no pushdown rule for the new plan node.

alex-spies replied:
>Yes, this is exactly what the code below at line 116 is guarding against.

It's not guarding against the hypothetical case where something drops the renamed columns, but then there's a `Project` later that "un-drops" them (even though that's a bug), as in `TopN BY field1, field2 | NewProjectionThatDropsRenamedColumns | ProjectThatUnDropsRenamedColumns | remote Enrich`.

But, since this is hypothetical, no need to be as defensive. This is fine.

smalyshev committed:
Make HoistTopN stop on topN

smalyshev committed:
fix test

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
fix test

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

alex-spies reviewed (approved):
This is great, thank you @smalyshev ! Finally, remote enrich will be working correctly and we don't have to work around it!

That was a seriously tricky one, and I'm really happy you made it work!

I have only some last, minor comments - let's :shipit: !

smalyshev committed:
Feedback

smalyshev committed:
Merge branch 'main' into entich-remote-unhack

smalyshev committed:
fix refactoring

smalyshev committed:
Merge branch 'main' into entich-remote-unhack
